agent/secrets/secrets.go

```diff
@@ -62,14 +62,26 @@ func (s *secrets) Reset() {
 type taskRestrictedSecretsProvider struct {
 	secrets   exec.SecretGetter
 	secretIDs map[string]struct{} // allow list of secret ids
+	taskID    string              // ID of the task the provider restricts for
 }
 
 func (sp *taskRestrictedSecretsProvider) Get(secretID string) (*api.Secret, error) {
 	if _, ok := sp.secretIDs[secretID]; !ok {
 		return nil, fmt.Errorf("task not authorized to access secret %s", secretID)
 	}
 
-	return sp.secrets.Get(secretID)
+	// First check if the secret is available with the task specific ID, which is the concatenation
+	// of the original secret ID and the task ID with a dot in between.
+	// That is the case when a secret driver has returned DoNotReuse == true for a secret value.
```

Comment:
For any kind of serialization, it's much clearer IMO to have both the serialization and deserialization functions right next to each other? Especially for internal serializations as it's the case here.

Reply:
> For any kind of serialization, it's much clearer IMO to have both the serialization and deserialization functions right next to each other? Especially for internal serializations as it's the case here.

Thanks for the feedback. I'm not sure I follow entirely; do you refer to `fmt.Sprintf("%s.%s", secretID, sp.taskID)`?

Reply:
Would you like it to be defined in a separate file or location? Remember, I don't have access to the secret *object* at the time, nor the task object.

Reply:
@wk8 Any chance you'll find time this week to elaborate? I'll be happy to make changes to make the PR more consistent and improve the code quality.

Reply:
@sirlatrom sorry for not replying sooner!

Yes, I do refer to `fmt.Sprintf("%s.%s", secretID, sp.taskID)`, ie the special secret ID format with a separating `.` for task-specific secrets.

Reply:
Cool! So does that mean it's good for merging now? @wk8 @justincormack @dperny @anshulpundir?